### PR TITLE
Implementation of additional Plugin API Methods

### PR DESCRIPTION
**Pull request by DSchalla**
This PR implements various API methods to provides plugins at least the same capabilities provided by the REST API as well as new hooks. 

**Hooks**
- [x] User Has Joined Channel
- [x] User Has Left Channel
- [x] User Has Joined Team
- [x] User Has Left Team
- [x] Channel Has Been Created
- [x] ...

**API Methods**
- [x] Team Member
- [x] Get Public Channel
- [x] Send Ephemeral Post
- [x] Config Management
- [x] ...

### CODE REVIEW

**Comment by DSchalla** [comment]
@crspeller @jwilander I've added a first implementation for the following hooks:

- User Has Joined Channel
- User Has Left Channel
- User Has Joined Team
- User Has Left Team

All hooks pass the team-/channel member object (even when it got deleted from the DB for consistency) and the actor user object for the person who actually executed the action. Feedback would be appreciated.

File: api4/channel_test.go [hunk 1]
```diff
@@ -1837,9 +1837,6 @@ func TestRemoveChannelMember(t *testing.T) {
 		t.Fatal("should have passed")
 	}
 
-	_, resp = Client.RemoveUserFromChannel(th.BasicChannel.Id, th.BasicUser2.Id)
-	CheckNoError(t, resp)
-
```
Comment (DSchalla) [inline api4/channel_test.go]: The same user / channel combination was attempted to deleted twice, which seems to be a mistake. The code before the introduced hook changes did not actually check whether a Channel Membership existed, I've removed the second test call in question, happy to discuss that of course.

**Comment by jwilander** [comment]
@DSchalla want us to review and get merged what you have here? Then if you want to add more, we can make another PR?

**Comment by DSchalla** [comment]
Sounds good to me!

File: app/channel.go [hunk 2]
```diff
@@ -1323,14 +1386,34 @@ func (a *App) RemoveUserFromChannel(userIdToRemove string, removerUserId string,
...
 
+	if a.PluginsReady() {
+		a.Go(func() {
+			pluginContext := &plugin.Context{}
+			a.Plugins.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+				hooks.UserHasLeftChannel(pluginContext, cm, actorUser)
+				return true
+			}, plugin.UserHasLeftChannelId)
+		})
+	}
```
Comment (jwilander) [inline app/channel.go]: It looks like both `RemoveUserFromChannel` and `LeaveChannel` call `removeUserFromChannel`. Might be able to move this down a level into that function so it's only in one place

File: app/team.go [hunk 3]
```diff
@@ -575,12 +592,21 @@ func (a *App) AddTeamMember(teamId, userId string) (*model.TeamMember, *model.Ap
...
 
+	if a.PluginsReady() {
+		a.Go(func() {
+			pluginContext := &plugin.Context{}
+			a.Plugins.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+				hooks.UserHasJoinedTeam(pluginContext, teamMember, nil)
+				return true
+			}, plugin.UserHasJoinedTeamId)
+		})
+	}
```
Comment (jwilander) [inline app/team.go]: Similar comment here (and actually I think this one calls `JoinUserToTeam` so this hook might be getting triggered twice and you might just be able to remove this one and it'll work

File: app/channel.go (hunk 2 above)
Comment (DSchalla) [reply app/channel.go]: Good call. I initially thought that this one doesnt pass the remover id, but that was apparently an oversight.

File: app/team.go (hunk 3 above)
Comment (DSchalla) [reply app/team.go]: Yup, just having it in `JoinUserToTeam` seems to be sufficient.

**Comment by DSchalla** [comment]
@jwilander Updated the PR to reflect your review, thanks!

**Pull request merged**